File: docs/external-access/read_compacted_state.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import os
import sys

import boto3


def env_required(name: str) -> str:
    value = os.getenv(name, "").strip()
    if not value:
        raise RuntimeError(f"Missing required env: {name}")
    return value


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Read QuantLab compacted state from S3.")
    p.add_argument("--exchange", default="")
    p.add_argument("--stream", default="")
    p.add_argument("--symbol", default="")
    p.add_argument("--date", default="")
    p.add_argument("--status", default="")
    p.add_argument("--day-quality", default="")
    p.add_argument("--limit", type=int, default=50)
    return p.parse_args()


def parse_partition_key(partition_key: str) -> dict[str, str] | None:
    parts = str(partition_key).split("/")
    if len(parts) != 4:
      return None
    exchange, stream, symbol, date = parts
    if len(date) != 8 or not date.isdigit():
        return None
    return {
        "exchange": exchange.lower(),
        "stream": stream.lower(),
        "symbol": symbol.lower(),
        "date": date,
    }


def read_state_from_s3() -> tuple[str, str, dict]:
    endpoint = env_required("S3_COMPACT_ENDPOINT")
    access_key = env_required("S3_COMPACT_ACCESS_KEY")
    secret_key = env_required("S3_COMPACT_SECRET_KEY")
    bucket = os.getenv("S3_COMPACT_BUCKET", "quantlab-compact")
    key = os.getenv("S3_COMPACT_STATE_KEY", "compacted/_state.json")
    region = os.getenv("S3_COMPACT_REGION", "us-east-1")

    s3 = boto3.client(
        "s3",
        endpoint_url=endpoint,
        aws_access_key_id=access_key,
        aws_secret_access_key=secret_key,
        region_name=region,
    )
    res = s3.get_object(Bucket=bucket, Key=key)
    return bucket, key, json.loads(res["Body"].read().decode("utf-8"))


def matches_filters(row: dict, args: argparse.Namespace) -> bool:
    if args.exchange and row["exchange"] != args.exchange.lower():
        return False
    if args.stream and row["stream"] != args.stream.lower():
        return False
    if args.symbol and row["symbol"] != args.symbol.lower():
        return False
    if args.date and row["date"] != args.date:
        return False
    if args.status and row["status"] != args.status.lower():
        return False
    if args.day_quality and row["day_quality_post"] != args.day_quality.upper():
        return False
    return True


def main() -> int:
    args = parse_args()
    bucket, key, state = read_state_from_s3()
    partitions = state.get("partitions")
    if not isinstance(partitions, dict):
        raise RuntimeError("state json missing partitions object")

    rows: list[dict] = []
    for partition_key, meta_raw in partitions.items():
        parsed = parse_partition_key(partition_key)
        if not parsed:
            continue
        meta = meta_raw if isinstance(meta_raw, dict) else {}
        row = {
            **parsed,
            "partition_key": partition_key,
            "status": str(meta.get("status", "")).strip().lower(),
            "day_quality_post": None if meta.get("day_quality_post") is None else str(meta.get("day_quality_post")).strip().upper(),
            "rows": meta.get("rows"),
            "total_size_bytes": meta.get("total_size_bytes"),
            "updated_at": meta.get("updated_at"),
        }
        row["data_key"] = f"exchange={row['exchange']}/stream={row['stream']}/symbol={row['symbol']}/date={row['date']}/data.parquet"
        row["meta_key"] = f"exchange={row['exchange']}/stream={row['stream']}/symbol={row['symbol']}/date={row['date']}/meta.json"
        row["data_s3_uri"] = f"s3://{bucket}/{row['data_key']}"
        row["meta_s3_uri"] = f"s3://{bucket}/{row['meta_key']}"
        if matches_filters(row, args):
            rows.append(row)

    rows.sort(key=lambda item: (item["date"], item["symbol"], item["stream"]))
    limit = args.limit if args.limit > 0 else 50
    payload = {
        "state_bucket": bucket,
        "state_key": key,
        "state_uri": f"s3://{bucket}/{key}",
        "last_compacted_date": state.get("last_compacted_date"),
        "updated_at": state.get("updated_at"),
        "match_count": len(rows),
        "results": rows[:limit],
    }
    json.dump(payload, sys.stdout, indent=2)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: docs/external-access/read-compacted-state.js
#!/usr/bin/env node

const { S3Client, GetObjectCommand } = require("@aws-sdk/client-s3");

function parseArgs(argv) {
  const out = { limit: 50 };
  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    const value = argv[i + 1];
    if (key === "--exchange") out.exchange = String(value || "").toLowerCase();
    if (key === "--stream") out.stream = String(value || "").toLowerCase();
    if (key === "--symbol") out.symbol = String(value || "").toLowerCase();
    if (key === "--date") out.date = String(value || "");
    if (key === "--status") out.status = String(value || "").toLowerCase();
    if (key === "--day-quality") out.dayQuality = String(value || "").toUpperCase();
    if (key === "--limit") out.limit = Number.parseInt(String(value || "50"), 10);
  }
  return out;
}

function envRequired(name) {
  const value = process.env[name];
  if (!value) throw new Error(`Missing required env: ${name}`);
  return value;
}

function parsePartitionKey(partitionKey) {
  const parts = String(partitionKey || "").split("/");
  if (parts.length !== 4) return null;
  const [exchange, stream, symbol, date] = parts;
  if (!/^\d{8}$/.test(date)) return null;
  return { exchange, stream, symbol, date };
}

async function readStateFromS3() {
  const endpoint = envRequired("S3_COMPACT_ENDPOINT");
  const accessKeyId = envRequired("S3_COMPACT_ACCESS_KEY");
  const secretAccessKey = envRequired("S3_COMPACT_SECRET_KEY");
  const bucket = process.env.S3_COMPACT_BUCKET || "quantlab-compact";
  const key = process.env.S3_COMPACT_STATE_KEY || "compacted/_state.json";
  const region = process.env.S3_COMPACT_REGION || "us-east-1";

  const s3 = new S3Client({
    endpoint,
    region,
    credentials: { accessKeyId, secretAccessKey },
    forcePathStyle: true
  });

  const res = await s3.send(new GetObjectCommand({ Bucket: bucket, Key: key }));
  const text = await res.Body.transformToString();
  return { bucket, key, state: JSON.parse(text) };
}

function matchesFilters(row, args) {
  if (args.exchange && row.exchange !== args.exchange) return false;
  if (args.stream && row.stream !== args.stream) return false;
  if (args.symbol && row.symbol !== args.symbol) return false;
  if (args.date && row.date !== args.date) return false;
  if (args.status && row.status !== args.status) return false;
  if (args.dayQuality && row.day_quality_post !== args.dayQuality) return false;
  return true;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const { bucket, key, state } = await readStateFromS3();
  if (!state || typeof state !== "object" || Array.isArray(state.partitions)) {
    throw new Error("Invalid state JSON");
  }

  const partitions = state.partitions;
  if (!partitions || typeof partitions !== "object") {
    throw new Error("state json missing partitions object");
  }

  const rows = [];
  for (const [partitionKey, metaRaw] of Object.entries(partitions)) {
    const parsed = parsePartitionKey(partitionKey);
    if (!parsed) continue;
    const meta = metaRaw && typeof metaRaw === "object" ? metaRaw : {};
    const row = {
      ...parsed,
      partition_key: partitionKey,
      status: String(meta.status || "").toLowerCase(),
      day_quality_post: meta.day_quality_post == null ? null : String(meta.day_quality_post).toUpperCase(),
      rows: meta.rows ?? null,
      total_size_bytes: meta.total_size_bytes ?? null,
      updated_at: meta.updated_at ?? null,
      data_key: `exchange=${parsed.exchange}/stream=${parsed.stream}/symbol=${parsed.symbol}/date=${parsed.date}/data.parquet`,
      meta_key: `exchange=${parsed.exchange}/stream=${parsed.stream}/symbol=${parsed.symbol}/date=${parsed.date}/meta.json`
    };
    row.data_s3_uri = `s3://${bucket}/${row.data_key}`;
    row.meta_s3_uri = `s3://${bucket}/${row.meta_key}`;
    if (matchesFilters(row, args)) rows.push(row);
  }

  rows.sort((a, b) => {
    return a.date.localeCompare(b.date) || a.symbol.localeCompare(b.symbol) || a.stream.localeCompare(b.stream);
  });

  const limit = Number.isFinite(args.limit) && args.limit > 0 ? args.limit : 50;
  console.log(JSON.stringify({
    state_bucket: bucket,
    state_key: key,
    state_uri: `s3://${bucket}/${key}`,
    last_compacted_date: state.last_compacted_date || null,
    updated_at: state.updated_at || null,
    match_count: rows.length,
    results: rows.slice(0, limit)
  }, null, 2));
}

main().catch((err) => {
  console.error(err.message || String(err));
  process.exit(1);
});
